quinn/src/connection.rs

```diff
@@ -27,44 +28,90 @@ use crate::{
 use proto::congestion::Controller;
 
 /// In-progress connection attempt future
+///
+/// On outgoing connections, this is given to the application in the state of the handshake already
+/// having begun. However, on incoming connections, this is given to the application before the
+/// server has begun performing its half of the handshake. This allows the application to check the
+/// remote address and reject the connection or have the client retry it with address validation
+/// before ever allocating state for it. Once this future is polled or other methods which take
+/// `&mut self` are called the server begins its half of the handshake and those methods can no
+/// longer be called.
 #[derive(Debug)]
 #[must_use = "futures/streams/sinks do nothing unless you `.await` or poll them"]
-pub struct Connecting {
+pub struct Connecting(Option<ConnectingState>);
```

Comment:
Why not expose `ConnectingState` directly?

Reply:
Responded below in thread.